tools/pubsys-config/src/vmware.rs

```diff
@@ -0,0 +1,231 @@
+const GOVC_NETWORK: &str = "GOVC_NETWORK";
+const GOVC_RESOURCE_POOL: &str = "GOVC_RESOURCE_POOL";
+const GOVC_FOLDER: &str = "GOVC_FOLDER";
+
+/// VMware-specific infrastructure configuration
+#[derive(Debug, Default, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct VmwareConfig {
+    #[serde(default)]
+    pub datacenters: Vec<String>,
+    #[serde(default)]
+    pub datacenter: HashMap<String, DatacenterBuilder>,
+    pub common: Option<DatacenterBuilder>,
+}
+
+/// VMware datacenter-specific configuration.
+///
+/// Fields are optional here because this struct is used to gather environment variables, common
+/// config, and datacenter-specific configuration, each of which may not have the complete set of
+/// fields.  It is used to build a complete datacenter configuration (hence the "Builder" name).
+#[derive(Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct DatacenterBuilder {
+    pub vsphere_url: Option<String>,
+    pub datacenter: Option<String>,
+    pub datastore: Option<String>,
+    pub network: Option<String>,
+    pub folder: Option<String>,
+    pub resource_pool: Option<String>,
+}
+
+/// Helper macro for retrieving a field from another struct if the field in `self` is `None`
+macro_rules! field_or {
+    ($self:expr, $field:ident, $other:expr) => {
+        $self
+            .$field
+            .as_ref()
+            .or($other.and_then(|o| o.$field.as_ref()))
+            .cloned()
+    };
+}
+
+impl DatacenterBuilder {
+    /// Create a DatacenterBuilder from environment variables
+    pub fn from_env() -> Self {
+        Self {
+            vsphere_url: get_env(GOVC_URL),
+            datacenter: get_env(GOVC_DATACENTER),
+            datastore: get_env(GOVC_DATASTORE),
+            network: get_env(GOVC_NETWORK),
+            folder: get_env(GOVC_FOLDER),
+            resource_pool: get_env(GOVC_RESOURCE_POOL),
+        }
+    }
+
+    /// Creates a new DatacenterBuilder, merging fields from another (Optional)
+    /// DatacenterBuilder if the field in `self` is None
+    pub fn take_missing_from(&self, other: Option<&Self>) -> Self {
+        Self {
+            vsphere_url: field_or!(self, vsphere_url, other),
+            datacenter: field_or!(self, datacenter, other),
+            datastore: field_or!(self, datastore, other),
+            network: field_or!(self, network, other),
+            folder: field_or!(self, folder, other),
+            resource_pool: field_or!(self, resource_pool, other),
+        }
+    }
+}
+
+/// A fully configured VMware datacenter, i.e. no optional fields
+#[derive(Debug)]
+pub struct Datacenter {
+    pub vsphere_url: String,
+    pub datacenter: String,
+    pub datastore: String,
+    pub network: String,
+    pub folder: String,
+    pub resource_pool: String,
+}
+
+/// Attempts to create a `Datacenter` from a `DatacenterBuilder` by ensuring that each
+/// field contains a value.
+impl TryFrom<DatacenterBuilder> for Datacenter {
```

Comment:
I think it might be more canonical for DatacenterBuilder to have:

```rust
    fn build(self) -> Result<Datacenter> {
        todo!();
    }
```

Reply:
I can see that.  Interested to hear some additional opinions.  It's not really a traditional builder per se, which is one of the reasons I leaned away from that.

Reply:
Looking at the call sites, I do think it's slightly confusing to implement the `try_from` trait for this purpose. It does seem like a very unique use-case.

I wonder if it's maybe simpler to just keep it to a single struct-type and just have a single `new` method that tries to create the config from multiple sources.

```
impl Datacenter {
    pub fn new(name: String) -> Result<Self> {
        // Try getting the config from environmental variables first
        todo!();
        // Try getting the config from infra.toml vmware config
        todo!();
        // Try getting the config from infra.toml vmware common config
        todo!();
        // If any config fields are still missing, return error MissingConfig
        todo!();
        
        Ok(
            Self {
                ...
            }
        )
}
```

Reply:
:)  This config piece is all a bit tricky and I spend a ton of time working through different ways to do it.  

What you suggested was my original idea but gathering the env/common multiple times (for multiple datacenters) seems a bit inefficient.  I suppose `Datacenter` could have a `new()` method that has arguments for env/common?  And we would call it for each datacenter in the config? I suppose in that case we'd end up using mostly the same logic as the `TryFrom` in order to get the final `Datacenter` since all the fields need to be present.

If most people are against the `TryFrom` implementation, I'm leaning towards the `build()` method as a solution.